add prometheus integration in recommendation-app

showuon opened the pull request:
This PR adds the prometheus integration in recommendation-app. This is for local environment demo use. Will add the integration with Openshift cluster built-in Prometheus examples in a separate PR.

showuon commented:
@katheris @tinaselenge , please take a look when available. Thanks.

katheris commented on recommendation-app/README.md:
````diff
@@ -57,4 +57,24 @@ The steps in the [README.md](../README.md) required to be completed for setting
+You can also deploy Prometheus to monitor the metrics inside job manager and task manager.
+5. Update the prometheus configuration with the ip address of flink pods to scrape from:
+   ```
+   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g recommendation-app/prometheus-install/prometheus-config.yaml
````
```suggestion
   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g recommendation-app/prometheus-install/prometheus-config.yaml
```

katheris commented on recommendation-app/README.md:
````diff
@@ -57,4 +57,24 @@ The steps in the [README.md](../README.md) required to be completed for setting
+5. Update the prometheus configuration with the ip address of flink pods to scrape from:
+   ```
+   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g recommendation-app/prometheus-install/prometheus-config.yaml
+   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g recommendation-app/prometheus-install/prometheus-config.yaml
````
```suggestion
   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g recommendation-app/prometheus-install/prometheus-config.yaml
```

katheris commented on recommendation-app/README.md:
````diff
@@ -57,4 +57,24 @@ The steps in the [README.md](../README.md) required to be completed for setting
+   ```
+   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g recommendation-app/prometheus-install/prometheus-config.yaml
+   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g recommendation-app/prometheus-install/prometheus-config.yaml
+   ```
````
These commands presumably work fine on linux, but not for Mac. So we should list these for linux and these for Mac:

```
sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g recommendation-app/prometheus-install/prometheus-config.yaml
sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g recommendation-app/prometheus-install/prometheus-config.yaml
```

showuon replied:
Ah, right, I was testing in Ubuntu. Added MacOS version. Thanks.

katheris commented on recommendation-app/README.md:
````diff
@@ -57,4 +57,24 @@ The steps in the [README.md](../README.md) required to be completed for setting
+
+7. Expose the prometheus UI with port-forward rule:
+   ```
+   kubectl port-forward svc/prometheus-service 9090
````
```suggestion
   kubectl port-forward svc/prometheus-service 9090 -n flink
```

katheris reviewed (commented):
I tried these and found a couple of updates were needed to work on Mac and also a few of the commands were missing `-n flink`.

More generally I wonder whether the instructions for deploying Prometheus should be in the `recommendation-app` README, or whether it would be better putting them in the top level README? Do you think we could write the instructions generically so if we added other examples in future it would explain how to use prometheus with any of the examples?

showuon commented:
Thanks for the review @katheris !

> I tried these and found a couple of updates were needed to work on Mac and also a few of the commands were missing -n flink.

Ah, right! I was testing in Ubuntu, and in my env, I set default namespace as `flink`, that's why I forgot to add `-n flink`. Sorry!

> More generally I wonder whether the instructions for deploying Prometheus should be in the recommendation-app README, or whether it would be better putting them in the top level README? Do you think we could write the instructions generically so if we added other examples in future it would explain how to use prometheus with any of the examples?

Good point! I've moved the `prometheus-install` into root folder, and reference it in `recommendation-app` readme. Let me know if you have other suggestion. Thank you.

katheris commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,32 @@
+# Integrate Prometheus into Flink cluster
+
+After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by these steps:
```
```suggestion
After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by following these steps:
```

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,32 @@
+
+2. Install prometheus, configuration, and service:
+   ```
+   kubectl apply -f . -n flink
````
```suggestion
   kubectl apply -f prometheus-install -n flink
```

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,32 @@
+
+   **Linux:**
+   ```
+   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml
````
```suggestion
   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
```

I think it's likely people would run this from the top level directory

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,32 @@
+   **Linux:**
+   ```
+   sed -i s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml
+   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-config.yaml
````
```suggestion
   sed -i s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
```

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,32 @@
+   ```
+   **MacOS**
+   ```
+   sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml
````
```suggestion
   sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-install/prometheus-config.yaml
```

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,32 @@
+   **MacOS**
+   ```
+   sed -i '' s/JOB_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f1)/g prometheus-config.yaml
+   sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-config.yaml
````
```suggestion
   sed -i '' s/TASK_MANAGER/$(kubectl get pods -lapp=recommendation-app -n flink -o=jsonpath="{range .items[*]}{.status.podIP}{','}{end}" | cut -d ',' -f2)/g prometheus-install/prometheus-config.yaml
```

katheris reviewed (approved):
I think we should assume people will run the commands from the top level directory so I've made some suggestions around that and a wording tweak. Otherwise looks good to me, thanks Luke

tinaselenge commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,46 @@
+
+2. Install prometheus, configuration, and service:
+   ```
+   kubectl apply -f prometheus-install -n flink
````
would this command apply everything under prometheus-install directory? if yes, I think it might be confusing because this directory has flink-monitor.yaml etc. Maybe we should have a sub directory for non openshift yamls only.

showuon replied:
Oops, you're right! Let me update it.

katheris commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,46 @@
+# Integrate Prometheus into Flink cluster in non-openshift cluster
```
```suggestion
# Integrate Prometheus into Flink cluster
```
I don't think we need the "in non-openshift cluster" part here since these are the instructions most people would follow

katheris commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,46 @@
+![img.png](job_metric.png)
+![img.png](task_metric.png)
+
+# Integrate Prometheus into Flink cluster in openshift cluster
```
```suggestion
# Integrate Prometheus into Flink cluster deployed on OpenShift
```

katheris commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,46 @@
+
+# Integrate Prometheus into Flink cluster in openshift cluster
+
+Since Openshift already has built-in Prometheus installed and configured, we can integrate with it by deploying `PodMonitor` CR for flink cluster:
```
```suggestion
Since OpenShift already has a built-in Prometheus installed and configured, we can integrate with it by deploying a `PodMonitor` CR for the flink cluster:
```

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,46 @@
+   ```
+   oc apply -f prometheus-install/flink-monitor.yaml -n flink
+   ```
+   Note: This CR is using `recommendation-app` as an example. Please update it if it's for other application.
````
```suggestion
   Note: This CR is configured to select the `FlinkDeployment` created as part of the `recommendation-app` example. Please update the `selector.matchLabels` field in `flink-monitor.yaml` if you are running a different example.

katheris commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,46 @@
+   ```
+   Note: This CR is using `recommendation-app` as an example. Please update it if it's for other application.
+
+2. It takes around 5 minutes to wait for prometheus operator to update the config for prometheus server. After that, you can query the metrics in the Openshift UI as described [here](https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-metrics.html#querying-metrics-for-all-projects-as-an-administrator_managing-metrics).
````
```suggestion
2. It takes around 5 minutes to wait for prometheus operator to update the config for prometheus server. After that, you can query the metrics in the OpenShift UI as described [here](https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-metrics.html#querying-metrics-for-all-projects-as-an-administrator_managing-metrics).
```

tinaselenge commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,46 @@
+# Integrate Prometheus into Flink cluster in non-openshift cluster
+
+After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by following steps:
```
```suggestion
After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics from job manager and task manager by following these steps:
```

tinaselenge commented on prometheus-install/README.md:
```diff
@@ -0,0 +1,46 @@
+
+After deploying Flink cluster, you can then deploy Prometheus to monitor the metrics inside job manager and task manager by following steps:
+
+1. Update the prometheus configuration with the ip address of flink pods to scrape from:
```
```suggestion
1. Update the prometheus configuration with the IP address of Flink pods to scrape from:
```

tinaselenge commented on prometheus-install/README.md:
````diff
@@ -0,0 +1,46 @@
+   ```
+   kubectl port-forward svc/prometheus-service -n flink 9090
+   ```
+4. Now you can monitor the metrics in job manager or task manager via the prometheus UI is accessible at localhost:9090.
````
```suggestion
4. Now you can monitor the metrics in job manager or task manager via the Prometheus UI is accessible at localhost:9090.
```

tinaselenge replied:
Prometheus and Flink sometimes starts with uppercase and sometimes with lowercase.

showuon commented:
@katheris @tinaselenge , thanks for the review. Addressed the comments. Thanks.

katheris reviewed (approved):
Thanks @showuon

kornys commented on prometheus-install/prometheus-server.yaml:
```diff
@@ -0,0 +1,27 @@
+    spec:
+      containers:
+        - name: prometheus
+          image: prom/prometheus
```
wouldn't be better to use quay path for prometheus image?
https://quay.io/repository/prometheus/prometheus

tinaselenge reviewed (approved):
Thanks Luke.

showuon commented:
@kornys , do you have any other comment?

kornys commented:
@showuon no, thank you for the fix and PR.
